finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/StreamProxy.scala

```diff
@@ -1,23 +1,84 @@
 package com.twitter.finagle.buoyant.h2
 import com.twitter.util.{Future, Try}
+import scala.collection.mutable
 
-/**
- * Wraps an underlying [[Stream]] with an [[StreamProxy.onFrame onFrame]] function.
- * The `onFrame` function is called for each frame in the stream.
- * @param underlying the [[Stream]] wrapped by this proxy
- * @param onFrame function called for each [[Frame]] in the underlying [[Stream]]
- */
-class StreamProxy(underlying: Stream, onFrame: Try[Frame] => Unit) extends Stream {
+abstract class StreamProxy(underlying: Stream) extends Stream {
+  /**
+   * @return true if the underlying stream is empty
+   */
   override def isEmpty: Boolean = underlying.isEmpty
-
-  override def read(): Future[Frame] = underlying.read().respond(onFrame)
-
   /**
    * Satisfied when an end-of-stream frame has been read from this
    * stream.
    *
    * If the stream is reset prematurely, onEnd fails with a [[Reset]].
    */
   override def onEnd: Future[Unit] = underlying.onEnd
+}
+/**
+ * Wraps an underlying [[Stream]] with an [[StreamOnFrame.onFrame onFrame]]
+ * function.
+ * The `onFrame` function is called for each frame in the stream.
+ *
+ * @param underlying the [[Stream]] wrapped by this proxy.
+ * @param onFrame function called for each [[Frame]] in the underlying [[Stream]].
+ */
+// TODO: consider renaming `onFrame` to `foreach`?
+class StreamOnFrame(underlying: Stream, onFrame: Try[Frame] => Unit)
+  extends StreamProxy(underlying) {
+  override def read(): Future[Frame] = underlying.read().respond(onFrame)
   override def toString: String = s"StreamProxy($underlying, onFrame=$onFrame)"
 }
+
+/**
+ * Wraps an underlying [[Stream]] with a function `f` that is called on each
+ * frame in the stream. The sequence of [[Frame]]s yielded by `f` will be
+ * inserted into the stream in order at the current position.
```

Comment:
Might reword this slightly to indicate that this returns a new stream which is backed by this stream with f applied on it.  imo "inserted" isn't really correct.